### Research on Factors affecting Rank:
- There are lot of factors affecting an apps rank for different keywords
- Evem though Google's actual factors for calculatig the rank is unknown, we can try to manipulate it to increase our ApP's visibility.
- This is where ASO comes in.
- Many factors like meta title, app name, URL,content relevancy, ratings, downloads etc. 

- From research we get to know that URL (i.e., App id name ) is one of the important factors for ASO rankings for keywords. <br>
Letss try to check it with our data.

In [4]:
import pandas as pd

In [5]:
df = pd.read_csv("A:/Next Labs test/browser_rankings_data.csv")

In [99]:
df.head(1)

ID  Keyword  Rank Country Language        Date                App ID  \
0   1  browser   2.0      in       en  12/16/2020  net.fast.web.browser   

  Date of Last Description Change  \
0                      12/18/2020   

                                   Short Description  \
0  The small, fastest, secure Web Browser, best m...   

                                    Long Description  
0  One of the smallest, fastest, secure Web Brows...

##### checking with value counts for data spread

In [109]:
len(df["ID"].value_counts())


2088

In [156]:
df["Keyword"].value_counts()

fast browser                    118
ad free browser                  93
android browser                  91
best privacy browsers            89
browser                          84
privacy browser                  74
privacy browsers for android     64
Name: Keyword, dtype: int64

In [155]:
(df["Rank"].value_counts())

2.0      23
5.0      23
4.0      20
7.0      20
6.0      20
8.0      19
3.0      19
1.0      17
14.0     17
10.0     17
9.0      16
11.0     13
12.0     11
13.0     10
15.0      9
20.0      9
16.0      8
17.0      8
24.0      8
26.0      7
18.0      7
28.0      7
21.0      6
23.0      6
42.0      6
22.0      6
29.0      6
27.0      6
34.0      6
19.0      6
39.0      6
31.0      6
40.0      6
105.0     5
52.0      5
38.0      5
30.0      5
43.0      5
44.0      5
48.0      5
81.0      4
53.0      4
36.0      4
89.0      4
67.0      4
79.0      4
51.0      4
32.0      4
35.0      4
25.0      4
78.0      4
33.0      4
69.0      4
60.0      4
58.0      4
65.0      4
50.0      4
54.0      3
68.0      3
73.0      3
94.0      3
61.0      3
97.0      3
100.0     3
107.0     3
83.0      3
106.0     3
87.0      3
86.0      3
104.0     3
90.0      3
75.0      3
71.0      3
96.0      3
74.0      3
82.0      3
109.0     3
55.0      3
63.0      3
37.0      3
59.0      3
57.0      3
45.0      3
116.

In [6]:
df["App ID"].value_counts()

com.duckduckgo.mobile.android    740
com.cloudmosa.puffinTV           608
com.brave.browser                370
com.vivaldi.browser              370
com.transsion.phoenix            370
com.opera.browser                304
net.fast.web.browser             152
com.android.chrome               152
Name: App ID, dtype: int64

In [153]:
df["Short Description"].value_counts()

Fast &amp; private web browser with ad blocker, sync, dark &amp; private mode.          198
Funny video, Football &amp; celebrity news, Fast &amp; data-saving browser.             125
Browse the web privately, search anonymously, and escape ad trackers.                    82
Puffin TV Browser - One App to Watch All Videos on the Web                               42
Puffin TV Browser - One App to Watch All Videos on the Web.                              42
Fast, safe web browser: free VPN, Ad blocker, QR-code reader &amp; personalized news     38
Free, safe internet browser! Private browsing, search engine &amp; pop-up blocker.       33
Fast, simple, and secure. Google Chrome browser for Android phones and tablets.          30
The small, fastest, secure Web Browser, best mobile browser, fast search Google.         23
Name: Short Description, dtype: int64

In [152]:
df["Long Description"].value_counts()

Vivaldi is a new private web browser with an integrated ad blocker, tracking protection, and useful built-in features. Browse with unique desktop-style tabs, end-to-end encrypted sync, dark & private mode. Download for free and browse fast, take notes in the browser, and capture full-length screenshots.\r\n\r\n⛔️ Block ads and trackers\r\nBlock ads and stop trackers from following you around the web to browse more privately and securely. Keep privacy ahead and say goodbye to annoying video ads, pop ups, and creepy trackers while browsing with Vivaldi.\r\n\r\nPlay with a flexible user interface\r\nGive your browser a look that's uniquely yours with Vivaldi’s flexible user interface (UI). Browse with dark/light modes using new desktop-style tabs, and keep go-to tools within easy reach.\r\n\r\n🕵️‍♂️ Private browsing\r\nKeep your browsing history discreet using incognito private tabs – searches, sites visited, cookies and temporary files won't be stored by Vivaldi. \r\n\r\n🛠 Manage tabs wi

In [128]:
df["Country"].value_counts()

in    3066
Name: Country, dtype: int64

In [129]:
df["Language"].value_counts()

en    3066
Name: Language, dtype: int64

In [131]:
len(df["Date"].value_counts())

76

In [132]:
len(df["Date of Last Description Change"].value_counts())

6

### Note that We have :
- 13 unique Long description
- 9 unique short description
- 8 unique App ID
- 119 different ranks with highly biased towards the first 15 ranks
- 7 unique keyword
- Interestingly 2088 unique ID's are found. Now since ID has no importance in understanding the data, Im gonna remove it.
- Also we remove other columns which has no significance to our problem statement from theory.
- Eg: country, Language is same, Dates have no signifucance either way for ranking.
- Now we check for duplicates in data.

In [145]:
df.drop(["ID","Date","Date of Last Description Change","Language","Country" ], axis=1, inplace = True)

In [146]:
duplicate = df[df.duplicated()]
len(duplicate) 

2453

- We can notice here that 2453 of the rows are just duplicates and has no significance on their presence to understand the ranking for ASO.
- So it shall be removed.

In [149]:
df = df.drop_duplicates()

In [150]:
len(df) # We will work on this 613 data for understanding our ranking factor.

613

In [51]:
print(df.columns.values)

['Keyword' 'Rank' 'App ID' 'Short Description' 'Long Description']


In [157]:
# Im going to save it into a csv for my easiness to view data.
#df.to_csv("A:\\Next Labs test\\preprocessed_browser_data.csv")

In [107]:
df = pd.read_csv("A:\\Next Labs test\\preprocessed_browser_data.csv",index_col = "Index")

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 613 entries, 0 to 3058
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Keyword            613 non-null    object 
 1   Rank               609 non-null    float64
 2   App ID             613 non-null    object 
 3   Short Description  613 non-null    object 
 4   Long Description   613 non-null    object 
dtypes: float64(1), object(4)
memory usage: 28.7+ KB


In [15]:
df.shape

(613, 5)

In [16]:
df.describe(include="all")

Keyword        Rank               App ID  \
count            613  609.000000                  613   
unique             7         NaN                    8   
top     fast browser         NaN  com.vivaldi.browser   
freq             118         NaN                  198   
mean             NaN   35.287356                  NaN   
std              NaN   33.107598                  NaN   
min              NaN    1.000000                  NaN   
25%              NaN    8.000000                  NaN   
50%              NaN   23.000000                  NaN   
75%              NaN   57.000000                  NaN   
max              NaN  134.000000                  NaN   

                                        Short Description  \
count                                                 613   
unique                                                  9   
top     Fast &amp; private web browser with ad blocker...   
freq                                                  198   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                         Long Description  
count                                                 613  
unique                                                 13  
top     Vivaldi is a new private web browser with an i...  
freq                                                  198  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN

In [18]:
df.isnull().sum() 

Keyword              0
Rank                 4
App ID               0
Short Description    0
Long Description     0
dtype: int64

- 4 null values found in Rank. Lets find them and see what we can do.
- We can use mode values of rank but since its just 4 values, we can remove those 4 rows.

In [108]:
null_data = df[df.isnull().any(axis=1)]
null_data

Keyword  Rank               App ID  \
Index                                                            
103           best privacy browsers   NaN  com.vivaldi.browser   
132                 privacy browser   NaN  com.vivaldi.browser   
1113                   fast browser   NaN  com.vivaldi.browser   
1760   privacy browsers for android   NaN  com.vivaldi.browser   

                                       Short Description  \
Index                                                      
103    Fast &amp; private web browser with ad blocker...   
132    Fast &amp; private web browser with ad blocker...   
1113   Fast &amp; private web browser with ad blocker...   
1760   Fast &amp; private web browser with ad blocker...   

                                        Long Description  
Index                                                     
103    Vivaldi is a new private web browser with an i...  
132    Vivaldi is a new private web browser with an i...  
1113   Vivaldi is a new private web browser with an i...  
1760   Vivaldi is a new private web browser with an i...

In [109]:
df = df.dropna(how='any',axis=0) 

In [110]:
df.isnull().sum()  

Keyword              0
Rank                 0
App ID               0
Short Description    0
Long Description     0
dtype: int64

In [111]:
df.shape

(609, 5)

 ***

In [113]:
df["App ID"].value_counts()

com.vivaldi.browser              194
com.transsion.phoenix            125
com.cloudmosa.puffinTV            84
com.duckduckgo.mobile.android     82
com.opera.browser                 38
com.brave.browser                 33
com.android.chrome                30
net.fast.web.browser              23
Name: App ID, dtype: int64

##### From App ID unique values, We can try to understand that the word "browser" being present in the App ID makes sense or what.

- We can use crosstab with ranks to understand relation with App ID. 
- Also we can make different small df with keywords so we can crosstab with rankings to understand them.

In [3]:
crosstb = pd.crosstab(df["App ID"], df["Keyword"])

NameError: name 'df' is not defined

In [ ]:
crosstb

##### From App ID values, the few categories makes sense is: || 
- 1. if keyword is present in the App ID
- 2. if keyword is not present in the App ID

 From the above crosstab, we an see that every keyword is contributing with <= 4 unique app ID's

In [63]:
android_browser	 = df[df['Keyword'] == "android browser"]
pd.crosstab(android_browser["App ID"], android_browser["Rank"])
# We can see that the keyword "android browser" has no impact in Ranks.

Rank                    2.0   3.0   4.0   5.0   6.0   7.0   8.0   9.0   10.0  \
App ID                                                                         
com.android.chrome         0     0     0     0     0     1     1     1     1   
com.cloudmosa.puffinTV     4     4     4     4     4     4     4     4     4   
com.opera.browser          2     2     2     2     2     2     2     2     0   
net.fast.web.browser       0     0     1     1     1     1     1     1     1   

Rank                    11.0  12.0  13.0  14.0  15.0  16.0  17.0  18.0  20.0  \
App ID                                                                         
com.android.chrome         1     1     1     1     1     1     1     1     1   
com.cloudmosa.puffinTV     4     4     0     4     0     0     0     0     0   
com.opera.browser          0     0     0     0     0     0     0     0     0   
net.fast.web.browser       1     1     1     1     1     1     0     0     0   

Rank                    24.0  
App ID                        
com.android.chrome         0  
com.cloudmosa.puffinTV     0  
com.opera.browser          0  
net.fast.web.browser       1

In [67]:
browser = df[df['Keyword'] == "browser"]
pd.crosstab(browser["App ID"], browser["Rank"])
# Keyword "browser" seems to have high relation but not.
# com.cloudmosa.puffinTV is showing better ranks of all which has no keyword in it.
# Also com.vivaldi.browser & com.brave.browser have no ranks  for "browser" keyword at all.

Rank                    1.0   2.0   3.0   4.0   5.0   6.0   7.0   8.0   9.0   \
App ID                                                                         
com.android.chrome         1     1     1     1     1     1     1     1     0   
com.cloudmosa.puffinTV     4     4     4     4     4     4     4     0     0   
com.opera.browser          2     2     2     2     2     2     2     2     2   
net.fast.web.browser       1     1     1     1     1     0     0     0     1   

Rank                    10.0  11.0  12.0  13.0  14.0  15.0  16.0  17.0  20.0  
App ID                                                                        
com.android.chrome         1     1     1     1     1     1     1     1     1  
com.cloudmosa.puffinTV     4     0     0     0     4     0     0     0     0  
com.opera.browser          2     0     0     0     2     0     0     0     0  
net.fast.web.browser       0     1     1     0     0     1     0     0     0

In [71]:
fast_browser = df[df['Keyword'] == "fast browser"]
pd.crosstab(fast_browser["App ID"], fast_browser["Rank"])
# Here net.fast.web.browser is not ranked at all when searched with keyword "fast browser"

Rank                           2.0    3.0    4.0    5.0    6.0    7.0    \
App ID                                                                    
com.brave.browser                  1      1      1      1      1      1   
com.duckduckgo.mobile.android      0      0      0      0      0      0   
com.transsion.phoenix              0      0      0      0      1      1   
com.vivaldi.browser                0      0      0      0      0      0   

Rank                           8.0    9.0    10.0   11.0   ...  101.0  103.0  \
App ID                                                     ...                 
com.brave.browser                  1      1      1      1  ...      0      0   
com.duckduckgo.mobile.android      0      0      0      2  ...      0      0   
com.transsion.phoenix              1      1      1      1  ...      0      0   
com.vivaldi.browser                0      0      0      0  ...      1      1   

Rank                           104.0  105.0  106.0  107.0  108.0  109.0  \
App ID                                                                    
com.brave.browser                  0      0      0      0      0      0   
com.duckduckgo.mobile.android      0      0      0      0      0      0   
com.transsion.phoenix              0      0      0      0      0      0   
com.vivaldi.browser                1      1      1      1      1      1   

Rank                           110.0  114.0  
App ID                                       
com.brave.browser                  0      0  
com.duckduckgo.mobile.android      0      0  
com.transsion.phoenix              0      0  
com.vivaldi.browser                1      1  

[4 rows x 76 columns]

In [73]:
privacy_browsers_for_android = df[df['Keyword'] == "privacy browsers for android"]
pd.crosstab(privacy_browsers_for_android["App ID"], privacy_browsers_for_android["Rank"])
# android broweser has some impact in its rank. But very less values(4) so cant consider.

Rank                           1.0    2.0    3.0    4.0    5.0    13.0   \
App ID                                                                    
com.brave.browser                  1      1      1      1      0      0   
com.duckduckgo.mobile.android      2      0      0      0      2      0   
com.transsion.phoenix              0      0      0      0      0      1   
com.vivaldi.browser                0      0      0      0      0      0   

Rank                           14.0   15.0   16.0   17.0   ...  86.0   87.0   \
App ID                                                     ...                 
com.brave.browser                  0      0      0      0  ...      0      0   
com.duckduckgo.mobile.android      0      0      0      0  ...      0      0   
com.transsion.phoenix              1      1      1      1  ...      0      0   
com.vivaldi.browser                0      0      0      0  ...      1      1   

Rank                           89.0   90.0   92.0   94.0   96.0   97.0   \
App ID                                                                    
com.brave.browser                  0      0      0      0      0      0   
com.duckduckgo.mobile.android      0      0      0      0      0      0   
com.transsion.phoenix              0      0      0      0      0      0   
com.vivaldi.browser                1      1      1      1      1      1   

Rank                           105.0  134.0  
App ID                                       
com.brave.browser                  0      0  
com.duckduckgo.mobile.android      0      0  
com.transsion.phoenix              0      0  
com.vivaldi.browser                1      1  

[4 rows x 60 columns]

In [74]:
# Conclusion about the URL is that: THe dataset is so random that there is no relation between the ranking and the App ID.

***

In [114]:
from functions import * # A text processing function I made.

In [115]:
df.columns

Index(['Keyword', 'Rank', 'App ID', 'Short Description', 'Long Description'], dtype='object')

In [132]:
df['clean_Short'] = df['Short Description'].map(lambda text: normalize_and_lemmaize(text))

In [133]:
df.isnull().sum()  

Keyword              0
Rank                 0
App ID               0
Short Description    0
Long Description     0
clean_Short          0
dtype: int64

In [134]:
df["clean_Short"].head(100)

Index
0      small fast secure web browser good mobile brow...
1      fast simple secure google chrome browser andro...
2      fast safe web browser free vpn ad blocker qrco...
3      fast safe web browser free vpn ad blocker qrco...
4              puffin tv browser one app watch video web
                             ...                        
140    funny video football amp celebrity news fast a...
142    fast amp private web browser ad blocker sync d...
145    funny video football amp celebrity news fast a...
147    fast amp private web browser ad blocker sync d...
150    funny video football amp celebrity news fast a...
Name: clean_Short, Length: 100, dtype: object

In [135]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [136]:
tfidf = TfidfVectorizer(ngram_range=(1,2),min_df=10)
data = tfidf.fit_transform(df['clean_Short'].values.astype('U'))

In [137]:
freq_df = pd.DataFrame(data.toarray(), columns=tfidf.get_feature_names_out())

In [138]:
df.tail()

Keyword  Rank                         App ID  \
Index                                                               
3035   best privacy browsers  44.0            com.vivaldi.browser   
3041            fast browser  16.0  com.duckduckgo.mobile.android   
3042            fast browser  16.0  com.duckduckgo.mobile.android   
3049         ad free browser  31.0          com.transsion.phoenix   
3058         android browser  15.0           net.fast.web.browser   

                                       Short Description  \
Index                                                      
3035   Fast &amp; private web browser with ad blocker...   
3041   Browse the web privately, search anonymously, ...   
3042   Browse the web privately, search anonymously, ...   
3049   Funny video, Football &amp; celebrity news, Fa...   
3058   The small, fastest, secure Web Browser, best m...   

                                        Long Description  \
Index                                                      
3035   Vivaldi is a new private web browser with an i...   
3041   Tired of being tracked online? We can help. At...   
3042   Tired of being tracked online? We can help. At...   
3049   Phoenix Browser is a fast, safe web browser fo...   
3058   One of the smallest, fastest, secure Web Brows...   

                                             clean_Short  
Index                                                     
3035   fast amp private web browser ad blocker sync d...  
3041   browse web privately search anonymously escape...  
3042   browse web privately search anonymously escape...  
3049   funny video football amp celebrity news fast a...  
3058   small fast secure web browser good mobile brow...

In [139]:
freq_df["browser"].tail(10)

599    0.000000
600    0.111710
601    0.111710
602    0.111710
603    0.111710
604    0.111710
605    0.000000
606    0.000000
607    0.112411
608    0.141063
Name: browser, dtype: float64

In [140]:
df["short_browser"] = freq_df["browser"]

In [141]:
df['clean_long'] = df['Long Description'].map(lambda text: normalize_and_lemmaize(text))

In [142]:
tfidf = TfidfVectorizer(ngram_range=(1,2),min_df=10)
data = tfidf.fit_transform(df['clean_long'].values.astype('U'))

In [143]:
freq_df1 = pd.DataFrame(data.toarray(), columns=tfidf.get_feature_names_out())

In [144]:
freq_df1["browser"]

0      0.101994
1      0.026979
2      0.072585
3      0.059159
4      0.105637
         ...   
604    0.237089
605    0.019574
606    0.023014
607    0.084202
608    0.101994
Name: browser, Length: 609, dtype: float64

In [145]:
df["long_browser"] = freq_df1["browser"]

In [146]:
len(df)

609

In [147]:
df["impact"] = df["short_browser"] > df["long_browser"]

In [153]:
df["impact"] = df.apply(lambda x: 1 if x["impact"] is True else 0, axis=1)
# 1 is for keyword in short description have more impact
# 0 fro keyword in long description have more impact

In [159]:
df["impact"].value_counts()

0    507
1    102
Name: impact, dtype: int64

In [161]:
pd.crosstab(df["impact"], df["Rank"])

Rank    1.0    2.0    3.0    4.0    5.0    6.0    7.0    8.0    9.0    10.0   \
impact                                                                         
0          13     14     11     11     16     18     17     17     12     15   
1           4      9      8      9      7      2      3      2      4      2   

Rank    ...  111.0  112.0  113.0  114.0  115.0  116.0  117.0  118.0  119.0  \
impact  ...                                                                  
0       ...      1      1      1      1      1      2      2      1      1   
1       ...      0      0      0      0      0      0      0      0      0   

Rank    134.0  
impact         
0           0  
1           1  

[2 rows x 119 columns]

- As we can see from the crosstab that, The number of ranks are quite high for Long Description.
- This goes valid for our dataset
- There for Long descriptions are having more correlation or impact in determining whether a Rank is 1 or 100.
- Repeativity of the keywords maybe a big factor for this result.


















































































































































##### Also there are only 13 unique Long descriptions and 9 unique short descriptions trying to values rank for 119 values.
##### I am gonna take a solid hunch from the data(browser rankings) I have seen that the above relation is not valid for real life datasets.
##### Same goes for.. whether a keyword present in first 10 words will make any impact in determining the rank of a browser app.


# Final Hypothesis

#### According to my research, the keywords in 6-15 words in the description is given more priority and weightage.
#### And the url or the App ID has some correlation in determining the rank of the app for that keyword.
#### And also precise Short description have more correlation than the long description.
#### These are the hypothesis I have reached after understanding Keyword rankings with SEO and ASO.